service/web/server.go

```diff
@@ -0,0 +1,66 @@
+package web
+
+import (
+	"context"
+	"regexp"
+
+	"github.com/razorpay/metro/internal/merror"
+	"github.com/razorpay/metro/pkg/logger"
+	metrov1 "github.com/razorpay/metro/rpc/proto/v1"
+)
+
+var projectIDRegex *regexp.Regexp
+
+func init() {
+	// Regex to capture project id from resource names
+	// Resources are in the form /projects/<project-id>/<resource-type>/<resource-id>
+	projectIDRegex = regexp.MustCompile(`^projects\/([^\/]+)\/.*`)
+}
+
+func getProjectIDFromRequest(ctx context.Context, req interface{}) (string, error) {
+	resourceName, err := getResourceNameFromRequest(ctx, req)
+	if err != nil {
+		return "", err
+	}
+	return getProjectIDFromResourceName(ctx, resourceName)
+}
+
+func getResourceNameFromRequest(ctx context.Context, req interface{}) (string, error) {
+	switch t := req.(type) {
```

Comment:
instead of request type, should we pass grpc method name and based on method name switch case, use appropriate type to desrialize

Reply:
We can do that. But a given proto request can be used by multiple methods. Hence using type switch will give us less switch cases